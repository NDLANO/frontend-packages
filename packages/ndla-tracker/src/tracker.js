/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-disable no-console */

const pageViewHistory = [];

function initializeGA(gaTrackingId) {
  // prettier-ignore
  window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)}; //eslint-disable-line
  window.ga.l = +new Date();
  window.ga('create', gaTrackingId, 'auto');
}

function resetDataLayer(googleTagManagerId) {
  if (window.dataLayer && window.google_tag_manager && googleTagManagerId) {
    window.google_tag_manager[googleTagManagerId].dataLayer.reset();
    window.dataLayer.push(window.originalLocation);
  }
}

export const configureTracker = ({
  listen,
  debug,
  gaTrackingId,
  googleTagManagerId,
}) => {
  initializeGA(gaTrackingId);
  // Push current page and start listning
  pageViewHistory.push({
    url: `${location.pathname}${location.search}${location.hash}`,
    tracked: false,
    debug,
  });

  listen(location => {
    if (debug) {
      console.info(
        `The current URL is ${location.pathname}${location.search}${location.hash}`,
      );
    }

    pageViewHistory.push({
      url: `${location.pathname}${location.search}${location.hash}`,
      tracked: false,
      debug,
      resetDataLayer: () => resetDataLayer(googleTagManagerId),
    });
  });
};

export const hasCurrentPageBeenTracked = () =>
  pageViewHistory[pageViewHistory.length - 1].tracked;

export const sendPageView = ({ title, dimensions }) => {
  const current = pageViewHistory[pageViewHistory.length - 1];
  current.tracked = true;

  if (current.debug) {
    console.info(`Tracking ${title} for page ${current.url}`);
    if (dimensions) {
      console.info('With dimensions: ', dimensions);
    }
  }

  // Always reset dataLayer before pushing new page events
  if (current.resetDataLayer) {
    current.resetDataLayer();
  }

  const dim = dimensions || { ga: {}, gtm: {} };

  window.ga('send', {
    hitType: 'pageview',
    page: current.url,
    title,
    ...dim.ga,
  });

  window.dataLayer.push({
    page_title: title,
    event: 'Pageview',
    url: current.url,
    ...dim.gtm,
  });
};
