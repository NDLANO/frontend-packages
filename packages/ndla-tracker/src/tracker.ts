/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { History } from 'history';

/* eslint-disable no-console */

declare global {
  interface Window {
    ga: any;
    dataLayer: any;
    google_tag_manager: any;
    originalLocation: unknown;
  }
}

type PageViewHistoryType = {
  url: string;
  tracked: boolean;
  debug: boolean;
  resetDataLayer?: () => void;
};

const pageViewHistory: PageViewHistoryType[] = [];

function initializeGA(gaTrackingId?: string) {
  window.ga =
    window.ga ||
    function () {
      // @ts-ignore
      (ga.q = ga.q || []).push(arguments);
    }; //eslint-disable-line
  window.ga.l = +new Date();
  window.ga('create', gaTrackingId, 'auto');
}

function resetDataLayer(googleTagManagerId?: string) {
  if (window.dataLayer && window.google_tag_manager && googleTagManagerId) {
    window.google_tag_manager[googleTagManagerId].dataLayer.reset();
    window.dataLayer.push(window.originalLocation);
  }
}

export type UnregisterCallback = () => void;

export const configureTracker = ({
  listen,
  debug,
  gaTrackingId,
  googleTagManagerId,
}: {
  listen: History<any>['listen'];
  debug?: boolean;
  gaTrackingId?: string;
  googleTagManagerId?: string;
}) => {
  initializeGA(gaTrackingId);
  // Push current page and start listning
  pageViewHistory.push({
    url: `${location.pathname}${location.search}${location.hash}`,
    tracked: false,
    debug: !!debug,
  });

  listen((location) => {
    if (debug) {
      console.info(`The current URL is ${location.pathname}${location.search}${location.hash}`);
    }

    pageViewHistory.push({
      url: `${location.pathname}${location.search}${location.hash}`,
      tracked: false,
      debug: !!debug,
      resetDataLayer: () => resetDataLayer(googleTagManagerId),
    });
  });
};

export const hasCurrentPageBeenTracked = () => pageViewHistory[pageViewHistory.length - 1].tracked;

export const sendPageView = ({
  title,
  dimensions,
}: {
  title: string;
  dimensions?: {
    ga: any;
    gtm: any;
  };
}) => {
  const current = pageViewHistory[pageViewHistory.length - 1];
  current.tracked = true;

  if (current.debug) {
    console.info(`Tracking ${title} for page ${current.url}`);
    if (dimensions) {
      console.info('With dimensions: ', dimensions);
    }
  }

  // Always reset dataLayer before pushing new page events
  if (current.resetDataLayer) {
    current.resetDataLayer();
  }

  const dim = dimensions || { ga: {}, gtm: {} };

  window.ga('send', {
    hitType: 'pageview',
    page: current.url,
    title,
    ...dim.ga,
  });

  window.dataLayer.push({
    page_title: title,
    event: 'Pageview',
    url: current.url,
    ...dim.gtm,
  });
};
