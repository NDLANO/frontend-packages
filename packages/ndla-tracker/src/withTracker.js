/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { Component } from 'react';
import hoistStatics from 'hoist-non-react-statics';
import { hasCurrentPageBeenTracked, sendPageView } from './tracker';

const mountedInstances = [];

const withTracker = WrappedComponent => {
  const Tracker = class extends Component {
    static trackPageView(props) {
      const lastMountedInstance = mountedInstances[mountedInstances.length - 1];

      if (hasCurrentPageBeenTracked() || lastMountedInstance !== WrappedComponent) {
        return;
      }

      const title = WrappedComponent.getDocumentTitle(props);
      const dimensions = WrappedComponent.getDimensions
        ? WrappedComponent.getDimensions(props)
        : undefined;
      sendPageView({ title, dimensions });
    }

    componentDidMount() {
      mountedInstances.push(WrappedComponent);

      if (!WrappedComponent.getDocumentTitle) {
        throw new Error(
          `Tracker expects a static getDocumentTitle function on the WrappedComponent.`,
        );
      }

      if (WrappedComponent.willTrackPageView) {
        WrappedComponent.willTrackPageView(Tracker.trackPageView, this.props);
      } else {
        Tracker.trackPageView(this.props);
      }
    }

    componentDidUpdate() {
      if (WrappedComponent.willTrackPageView) {
        WrappedComponent.willTrackPageView(Tracker.trackPageView, this.props);
      } else {
        Tracker.trackPageView(this.props);
      }
    }

    componentWillUnmount() {
      const index = mountedInstances.indexOf(WrappedComponent);
      mountedInstances.splice(index, 1);
    }

    render() {
      return <WrappedComponent {...this.props} trackPageView={Tracker.trackPageView} />;
    }
  };

  return hoistStatics(Tracker, WrappedComponent);
};

export default withTracker;
