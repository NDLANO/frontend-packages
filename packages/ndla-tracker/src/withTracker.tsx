/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { Component } from 'react';
import hoistStatics from 'hoist-non-react-statics';
import { hasCurrentPageBeenTracked, sendPageView } from './tracker';

const mountedInstances: React.ComponentType<any>[] = [];

type TrackableComponent<P> = React.ComponentType<P> & {
  getDocumentTitle: (p: P) => string;
  getDimensions?: (p: P) => { ga: any; gtm: any };
  willTrackPageView?: (callback: (p: P) => void, p: P) => void;
};

function withTracker<P>(WrappedComponent: TrackableComponent<P>): React.ComponentType<P> {
  const Tracker = class extends Component<P> {
    static trackPageView(props: P) {
      const lastMountedInstance = mountedInstances[mountedInstances.length - 1];

      if (hasCurrentPageBeenTracked() || lastMountedInstance !== WrappedComponent) {
        return;
      }

      const title = WrappedComponent.getDocumentTitle(props);
      const dimensions = WrappedComponent.getDimensions ? WrappedComponent.getDimensions(props) : undefined;
      sendPageView({ title, dimensions });
    }

    componentDidMount() {
      mountedInstances.push(WrappedComponent);

      // Kept for js interop, but should not be hit in ts.
      if (!WrappedComponent.getDocumentTitle) {
        throw new Error(`Tracker expects a static getDocumentTitle function on the WrappedComponent.`);
      }

      if (WrappedComponent.willTrackPageView) {
        WrappedComponent.willTrackPageView(Tracker.trackPageView, this.props);
      } else {
        Tracker.trackPageView(this.props);
      }
    }

    componentDidUpdate() {
      if (WrappedComponent.willTrackPageView) {
        WrappedComponent.willTrackPageView(Tracker.trackPageView, this.props);
      } else {
        Tracker.trackPageView(this.props);
      }
    }

    componentWillUnmount() {
      const index = mountedInstances.indexOf(WrappedComponent);
      mountedInstances.splice(index, 1);
    }

    render() {
      return <WrappedComponent {...this.props} trackPageView={Tracker.trackPageView} />;
    }
  };

  return hoistStatics(Tracker, WrappedComponent);
}

export default withTracker;
