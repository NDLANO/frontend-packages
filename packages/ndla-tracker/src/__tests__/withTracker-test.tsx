/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-env jest */
/* eslint-disable react/no-multi-comp */
import React, { Component, ReactNode } from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import sinon from 'sinon';
import withTracker from '../withTracker';
import * as tracker from '../tracker';

window.dataLayer = [];
window._mtm = [];
const history = createMemoryHistory();
tracker.configureTracker({ listen: history.listen });

interface PageProps {
  world?: string;
  children?: ReactNode;
}

class Page extends Component<PageProps> {
  static getDocumentTitle(props: PageProps) {
    return `Hello ${props.world}`;
  }

  render() {
    return (
      <div>
        <h1>Main Page</h1>
        {this.props.children}
      </div>
    );
  }
}

interface ChildPageProps {
  world: string;
}

class ChildPage extends Component<ChildPageProps> {
  static getDocumentTitle(props: ChildPageProps) {
    return `Hello ${props.world} from child page`;
  }

  render() {
    return (
      <div>
        <h1>Child Page</h1>
      </div>
    );
  }
}

test('withTracker HOC renderers Page correctly', () => {
  const PageWithTracker = withTracker(Page);
  const { container } = render(<PageWithTracker />);

  expect(container.firstChild).toMatchSnapshot();
});

test('sendPageView is called on component mounth ', () => {
  const PageWithTracker = withTracker(Page);
  const spy = sinon.spy(tracker, 'sendPageView');

  history.push('/test');
  render(<PageWithTracker world="world" />);

  expect(spy.calledOnce).toBe(true);
  expect(spy.args[0]).toMatchSnapshot();
  spy.restore();
});

test('sendPageView is called on component update if url is untracked', () => {
  const PageWithTracker = withTracker(Page);
  const spy = sinon.spy(tracker, 'sendPageView');

  history.push('/test1');
  const { rerender } = render(<PageWithTracker world="world" />);
  history.push('/test2');
  rerender(<PageWithTracker world="world" />);

  expect(spy.calledTwice).toBe(true);
  expect(spy.args).toMatchSnapshot();
  spy.restore();
});

test('sendPageView is not called on component update if url is tracked', () => {
  const PageWithTracker = withTracker(Page);
  const spy = sinon.spy(tracker, 'sendPageView');

  history.push('/test1');
  const { rerender } = render(<PageWithTracker world="world" />);
  rerender(<PageWithTracker world="world" />);

  expect(spy.calledOnce).toBe(true);
  expect(spy.args[0]).toMatchSnapshot();
  spy.restore();
});

test('sendPageView is called from child component if two page tracking compontents are in the same hierarchy', () => {
  const PageWithTracker = withTracker(Page);
  const ChildPageWithTracker = withTracker(ChildPage);
  const spy = sinon.spy(tracker, 'sendPageView');

  history.push('/test');
  render(
    <PageWithTracker world="world">
      <ChildPageWithTracker world="world" />
    </PageWithTracker>,
  );

  expect(spy.calledOnce).toBe(true);
  expect(spy.args[0]).toMatchSnapshot();
  spy.restore();
});
