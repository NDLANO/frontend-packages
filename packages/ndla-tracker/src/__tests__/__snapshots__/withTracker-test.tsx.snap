// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendPageView is called from child component if two page tracking compontents are in the same hierarchy 1`] = `
Array [
  Object {
    "dimensions": undefined,
    "title": "Hello world from child page",
  },
]
`;

exports[`sendPageView is called on component mounth  1`] = `
Array [
  Object {
    "dimensions": undefined,
    "title": "Hello world",
  },
]
`;

exports[`sendPageView is called on component update if url is untracked 1`] = `
Array [
  Array [
    Object {
      "dimensions": undefined,
      "title": "Hello world",
    },
  ],
  Array [
    Object {
      "dimensions": undefined,
      "title": "Hello world",
    },
  ],
]
`;

exports[`sendPageView is not called on component update if url is tracked 1`] = `
Array [
  Object {
    "dimensions": undefined,
    "title": "Hello world",
  },
]
`;

exports[`withTracker HOC renderers Page correctly 1`] = `
<div>
  <h1>
    Main Page
  </h1>
</div>
`;
